contracts/dtoken/src/user_flow_protection.rs

```diff
@@ -0,0 +1,91 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::{AccountId, BlockHeight, log, near_bindgen};
+use near_sdk::collections::{UnorderedMap};
+use near_sdk::env::block_height;
+
+const BLOCKS_TO_NEXT_OPERATION: BlockHeight = 100;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct UserPerform {
+    last_user_perform: UnorderedMap<AccountId, BlockHeight>,
+}
+
+impl Default for UserPerform {
+    fn default() -> Self {
+        Self {
+            last_user_perform: UnorderedMap::new(b"s".to_vec()),
+        }
+    }
+}
+
+#[near_bindgen]
+impl UserPerform {
+    pub fn set_last_user_perform(&mut self, account_id: AccountId) {
+        let block_height = block_height();
+        log!("Set last perform account: {} block_height: {}", account_id, block_height);
+        self.last_user_perform.insert(&account_id, &block_height);
+    }
+
+    pub fn get_last_user_perform(&self, account_id: AccountId) -> Option<BlockHeight> {
+        log!("Get last perform account: {}", account_id);
+        self.last_user_perform.get(&account_id)
+    }
+
+    pub fn is_user_can_perform_operation(&self, account_id: AccountId) -> bool {
+        log!("User can perform: {}", account_id);
+        let last_perform_block_height = self.get_last_user_perform(account_id);
+        let current_block_height = block_height();
+        current_block_height - last_perform_block_height.unwrap() >= BLOCKS_TO_NEXT_OPERATION
```

Comment:
From this logic, the user can perform 1 action per 100 blocks? As I remember, we said that if the user can perform the action we should set a flag, that the user is blocked. While the user is blocked we don't allow him to do any operation, except the situation when he was blocked more then BLOCKS_TO_NEXT_OPERATION blocks